### PR TITLE
feat: global search

### PR DESCRIPTION
**Pull request by reallybeard**
## Description

- This adds a new global search to the header
- Adds a new single transaction page

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #

## Risk

mild risk

### Operations

Test that you can search and find assets, opportunities, tx history. And clicking on them goes to the right place.

## Screenshots (if applicable)
Screenshot 2023-05-03 at 12 07 34 PM

### CODE REVIEW

File: src/components/Layout/Header/GlobalSearch/StakingResults/StakingResult.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  activeIndex?: number
+  onClick: (arg: GlobalSearchResult) => void
+}
+
+export const StakingResult = forwardRef<StakingResultProps, 'div'>(
+  ({ stakingId, onClick, index, activeIndex }, ref) => {
+    const stakingOpportunities = useAppSelector(
+      selectAggregatedEarnUserStakingOpportunitiesIncludeEmpty,
+    )
+    const opportunity = stakingOpportunities.find(opportunity => opportunity.id === stakingId)
```
Comment (0xdef1cafe) [inline src/components/Layout/Header/GlobalSearch/StakingResults/StakingResult.tsx:30]: this is going to return a new ref every time - should be able to select by staking id

File: src/components/Layout/Header/GlobalSearch/utils.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import qs from 'qs'
+import { trackOpportunityEvent } from 'lib/mixpanel/helpers'
+import { MixPanelEvents } from 'lib/mixpanel/types'
+import type {
+  LpEarnOpportunityType,
+  OpportunityId,
+  StakingEarnOpportunityType,
+} from 'state/slices/opportunitiesSlice/types'
+
+type goToOpportunityProps = {
```
Comment (0xdef1cafe) [inline src/components/Layout/Header/GlobalSearch/utils.ts:16]:
```suggestion
type GoToOpportunityProps = {
```

File: src/components/Layout/Header/GlobalSearch/utils.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  stakingOpportunities: StakingEarnOpportunityType[]
+  lpOpportunities: LpEarnOpportunityType[]
+  opportunityId: OpportunityId
+  opportunityType: DefiType
+  action: DefiAction
+  location: Location
+  assets: Partial<Record<AssetId, Asset>>
+}
+export const GoToOpportunity = ({
```
Comment (0xdef1cafe) [inline src/components/Layout/Header/GlobalSearch/utils.ts:25]: `makeOpportunityRouteDetails`?

File: src/state/slices/opportunitiesSlice/selectors/combined.ts [hunk 4]
```diff
@@ -460,7 +460,7 @@ export const selectAggregatedEarnOpportunitiesByProvider = createDeepEqualOutput
         acc[provider].opportunities.lp.push(cur.id)
       }
 
-      if (isActiveOpportunityByFilter) {
+      if (cur.type === DefiType.Staking && isActiveOpportunityByFilter) {
```
Comment (0xdef1cafe) [inline src/state/slices/opportunitiesSlice/selectors/combined.ts:463]: ?

File: src/state/slices/search-selectors.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      .map(asset => {
+        return {
+          type: GlobalSearchResultType.Asset,
+          id: asset.assetId,
+        }
+      })
+      .slice(0, limit)
+    const opportunities = opportunityResults.reduce(
+      (acc: { staking: string[]; lp: string[] }, curr: AggregatedOpportunitiesByProviderReturn) => {
+        acc['staking'].push(...curr.opportunities.staking)
```
Comment (0xdef1cafe) [inline src/state/slices/search-selectors.ts:47]: don't spread for perf

File: src/state/slices/txHistorySlice/selectors.ts [hunk 6]
```diff
@@ -243,3 +247,28 @@ export const selectMaybeNextAccountNumberByChainId = createSelector(
     return [isAbleToAddNextAccount, isAbleToAddNextAccount ? nextAccountNumber : null]
   },
 )
+export const SelectTxsByQuery = createDeepEqualOutputSelector(
+  selectTxs,
+  selectAssets,
+  selectSearchQueryFromFilter,
+  (txsById, assets, searchQuery) => {
+    const txArray: [string, Tx][] = Object.entries(txsById)
```
Comment (0xdef1cafe) [inline src/state/slices/txHistorySlice/selectors.ts:255]:
```suggestion
    const txArray: [TxId, Tx][] = Object.entries(txsById)
```

File: src/state/slices/txHistorySlice/selectors.ts [hunk 7]
```diff
@@ -243,3 +247,28 @@ export const selectMaybeNextAccountNumberByChainId = createSelector(
...
+    if (!searchQuery)
+      return Object.entries(txsById)
+        .sort((a, b) => b[1].blockTime - a[1].blockTime)
```
Comment (0xdef1cafe) [inline src/state/slices/txHistorySlice/selectors.ts:258]: we can avoid this by sorting `txsById` going into the store

File: src/state/slices/txHistorySlice/selectors.ts [hunk 8]
```diff
@@ -243,3 +247,28 @@ export const selectMaybeNextAccountNumberByChainId = createSelector(
     return [isAbleToAddNextAccount, isAbleToAddNextAccount ? nextAccountNumber : null]
   },
 )
+export const SelectTxsByQuery = createDeepEqualOutputSelector(
```
Comment (0xdef1cafe) [inline src/state/slices/txHistorySlice/selectors.ts:250]:
```suggestion
export const selectTxsByQuery = createDeepEqualOutputSelector(
```

File: src/components/Layout/Header/GlobalSearch/utils.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  opportunityId,
+  action,
+  opportunityType,
+  location,
+  assets,
+}: goToOpportunityProps) => {
+  const lpOpportunity = lpOpportunities.find(lpOpportunity => lpOpportunity.id === opportunityId)
+  const stakingOpportunity = stakingOpportunities.find(
+    stakingOpportunity => stakingOpportunity.id === opportunityId,
+  )
```
[the remark this answers is not in the document]
Comment (reallybeard) [reply src/components/Layout/Header/GlobalSearch/utils.ts:37]: we can't use a hook within this so open to feedback on another way to do this @0xdef1cafe

[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/components/Layout/Header/GlobalSearch/utils.ts:37]: oh yeah this is a util we're ok

File: src/components/Layout/Header/GlobalSearch/StakingResults/StakingResult.tsx [hunk 10]
```diff
@@ -24,10 +24,9 @@ type StakingResultProps = {
 
 export const StakingResult = forwardRef<StakingResultProps, 'div'>(
   ({ stakingId, onClick, index, activeIndex }, ref) => {
-    const stakingOpportunities = useAppSelector(
-      selectAggregatedEarnUserStakingOpportunitiesIncludeEmpty,
+    const opportunity = useAppSelector(state =>
+      selectAggregatedEarnUserStakingOpportunitiesIncludeEmptyByStakingId(state, { stakingId }),
```
Comment (0xdef1cafe) [inline src/components/Layout/Header/GlobalSearch/StakingResults/StakingResult.tsx:28]: memo this `{ stakingId }` and you get a juicy stamp

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**